File: project/src/main/java/com/example/project/dto/ProjectDto.java
package com.example.project.dto;

import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.List;

public class ProjectDto {

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class ProjectSummary {
        private Long id;
        private String name;
        private String subjectName;
        private String status;
        private String role;
        private LocalDateTime createdAt;
    }

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class TeamMemberInfo {
        private Long id;
        private Long userId;
        private String firstName;
        private String lastName;
        private String role;
    }

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class ProjectDetails {
        private Long id;
        private String name;
        private String description;
        private String subjectName;
        private String status;
        private LocalDateTime createdAt;
        private List<TeamMemberInfo> members;
        private String myRole;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class CreateRequest {
        @NotBlank(message = "Название проекта обязательно")
        @Size(min = 3, max = 255, message = "Название от 3 до 255 символов")
        private String name;

        @NotBlank(message = "Описание проекта обязательно")
        @Size(max = 1500, message = "Описание не более 1500 символов")
        private String description;

        @NotBlank(message = "Предмет обязателен")
        @Size(min = 2, max = 255, message = "Предмет от 2 до 255 символов")
        private String subjectName;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class UpdateRequest {
        @NotBlank(message = "Название проекта обязательно")
        @Size(min = 3, max = 255, message = "Название от 3 до 255 символов")
        private String name;

        @NotBlank(message = "Описание проекта обязательно")
        @Size(max = 2000, message = "Описание не более 2000 символов")
        private String description;

        @NotBlank(message = "Предмет обязателен")
        @Size(min = 2, max = 255, message = "Предмет от 2 до 255 символов")
        private String subjectName;

        @NotBlank(message = "Статус обязателен")
        private String status;
    }

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class TaskTypeStats {
        private long total;
        private long closed;
        private long overdue;
    }

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class TaskStats {
        private TaskTypeStats assigned; // назначенные мне
        private TaskTypeStats all;      // все в проекте
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class AddMemberRequest {
        @NotNull(message = "ID пользователя обязателен")
        private Long userId;

        @Size(max = 100, message = "Роль не более 100 символов")
        private String role;
    }
}

File: front/src/dto/projectDto.js
// DTO-фабрики для запросов и ответов, связанных с проектами
export function CreateProjectDto(name, description, subjectName, deadline) {
    return { name, description, subjectName, deadline };
}

export function UpdateProjectDto(name, description, subjectName, status, deadline) {
    return { name, description, subjectName, status, deadline };
}

export function AddMemberDto(userId, role) {
    return { userId, role };
}

export function ProjectSummaryDto(data) {
    if (!data) return null;
    return {
        id: data.id,
        name: data.name,
        subjectName: data.subjectName,
        status: data.status,
        role: data.role,
        createdAt: data.createdAt,
        deadline: data.deadline,
    };
}

export function TeamMemberInfoDto(data) {
    if (!data) return null;
    return {
        id: data.id,
        userId: data.userId,
        firstName: data.firstName,
        lastName: data.lastName,
        role: data.role,
    };
}

export function ProjectDetailsDto(data) {
    if (!data) return null;
    return {
        id: data.id,
        name: data.name,
        description: data.description,
        subjectName: data.subjectName,
        status: data.status,
        createdAt: data.createdAt,
        deadline: data.deadline,
        members: Array.isArray(data.members) ? data.members.map(TeamMemberInfoDto) : [],
        myRole: data.myRole,
    };
}

export function TaskTypeStatsDto(data) {
    if (!data) return { total: 0, closed: 0, overdue: 0 };
    return {
        total: data.total || 0,
        closed: data.closed || 0,
        overdue: data.overdue || 0,
    };
}

export function TaskStatsDto(data) {
    if (!data) return {
        assigned: { total: 0, closed: 0, overdue: 0 },
        all: { total: 0, closed: 0, overdue: 0 }
    };
    return {
        assigned: TaskTypeStatsDto(data.assigned),
        all: TaskTypeStatsDto(data.all),
    };
}
